docs/proposals/11-cluster-health.md

```diff
@@ -0,0 +1,298 @@
+- Alternatively, patch the `kube-controller-manager` `Deployment` to disable the `nodelifecycle` controller.
+
+#### ETCDReachable
+
+The `Endpoint` monitoring part of the `root` [command](https://github.com/gardener/dependency-watchdog/blob/master/cmd/root.go) of the [dependency-watchdog](https://github.com/gardener/dependency-watchdog) can be re-factored into a separate component that updates the conditions with the `ETCDReachable` condition type.
+
+##### Possible Actions
+
+- When this condition transitions from status `false` to `true` (i.e. `etcd` transitions from being unreachable to reachable), the dependent `pods` could be deleted but only if they are in `CrashLoopBackoff`. This is to make sure the dependent `pods` recover as fast as possible from an `etcd` down time. This could be done by the restarter part of the current `root` [command](https://github.com/gardener/dependency-watchdog/blob/master/cmd/root.go) of the [dependency-watchdog](https://github.com/gardener/dependency-watchdog) but re-factored as a separate restarter component which restarts (deletes) some targets `pods` based on some configured (via command-line argument, environment variables, config files etc.) conditions in the `Cluster` resource but only if the `pods` are in `CrashLoopBackoff`. 
+
+#### KubeAPIServerReachableInternally
+
+The `KubeAPIServerReachableInternally` condition type can be implemented (both the monitoring and the remedial action parts) similar to the [`ETCDReachable`](#etcdreachable) condition type.
+
+This condition may not be directly usable to take action.
+However, it might be useful in avoiding taking some pointless (and potentially harmful) actions.
+For example, if `KubeAPIServerReachableInternally` is `false` then we can avoid [scaling down the `kube-controller-manager`](#possible-actions) even if `KubeAPIServerReachableExternally` is `false`.
+
+#### UnhealthyNodesThresholdReached
+
+A component can monitor the fraction of the `NotReady` nodes in a zone and update the conditions with the `UnhealthyNodesThresholdReached` condition type.
+The list of affected zones could be maintained in the [`Message`](https://github.com/kubernetes/cluster-registry/blob/master/pkg/apis/clusterregistry/v1alpha1/types.go#L168).
+
+##### Possible Actions
+
+- If necessary, `kube-controller-manager` could be scaled down as described [above](#possible-actions).
+- Some resource creations (say, `pods`) could be throttled/rate-limited.
+- The `machine-controller-manager` could be either scaled down as described [above](#possible-actions) or be enhanced to avoid (or ramp down) deletion of machines from the affected zones (or `machinedeployments`).
```

Comment:
With the extensibility, machine-controller-manager is just an internal component used by the extension controllers. How the remedy component decides that there is a machine-controller-manager? The contract currently is that gardenlet deploys a Worker resource, and it is up to the extension controller to create Nodes out of provided Worker spec.

Reply:
> With the extensibility, machine-controller-manager is just an internal component used by the extension controllers.

Yes. And also `machine-controller-manager` is an *external* component because it might be (probably already is being) used independent of gardener.

> How the remedy component decides that there is a machine-controller-manager?

Remedy component itself might be an extension or some hidden component which is deployed only if `machine-deployment-manager` is used in the seeds.

The deploy-time and runtime dependencies between extensions and internal components in the seed is a different problem and currently solved implicitly (by hard-coding or relying on the landscape admins to do the *right thing*). If this problem gets big enough, we might need a declarative solution for it.

Reply:
> Yes. And also machine-controller-manager is an external component because it might be (probably already is being) used independent of gardener.

Yes, it can be used independently.

> Remedy component itself might be an extension or some hidden component which is deployed only if machine-deployment-manager is used in the seeds.

This would mean that you bind a new (probable) extension (remedy component) with the knowledge that some other extension may (or may not) use machine-controller-manager. Currently the contract is on Worker level.
I am interested to see how this GEP can be adapted having in mind `GEP-01: Extensibility`.

> The deploy-time and runtime dependencies between extensions and internal components in the seed is a different problem and currently solved implicitly (by hard-coding or relying on the landscape admins to do the right thing)

WDYM? Which issue is solved by hard-coding or relying on the landscape admins to do the right thing?

Reply:
> This would mean that you bind a new (probable) extension (remedy component) with the knowledge that some other extension may (or may not) use machine-controller-manager. 
Currently the contract is on Worker level.
I am interested to see how this GEP can be adapted having in mind GEP-01: Extensibility.

This proposal is not in violation of [GEP-01](https://github.com/gardener/gardener/blob/master/docs/proposals/01-extensibility.md#proposal). In fact, it takes the spirit of that proposal forward.

I quote from [GEP-01]( https://github.com/gardener/gardener/blob/master/docs/proposals/01-extensibility.md#proposal).
```
This way one complex strongly coupled Gardener implementation covering all infrastructures is decomposed into a set of loosely coupled controllers implementing aspects of APIs defined by Gardener.
```
Loose coupling is not *no coupling*. I am proposing exactly such loose coupling. A remedy controller doesn't need to know about machine-controller-manager. All it needs to know is some component is maintaining the relevant cluster health conditions that it is interested in. 

Having said that, my other point in the previous comment was that there are multiple implicit dependencies between different extensions and components. Please see below for the examples.

> WDYM? Which issue is solved by hard-coding or relying on the landscape admins to do the right thing?

There are many such implicit dependencies that are either hard-coded or relies on the landscape admins to deploy the right combinations of extensions.

- `StatefulSet` deployment implicitly depends on backup infrastructure. This is carried forward by etcd-druid too.
- Cluster Autoscaler is a custom fork with machine-controller-manager integration. So, it expects that machine-controller-manager is going to be used.

These are just examples. I am sure we will find others if we dig further. [GEP-01](https://github.com/gardener/gardener/blob/master/docs/proposals/01-extensibility.md) is silent on such dependencies between extensions and controllers. This might be fine for now. But eventually we might need a good way to make these dependencies explicit and/or more loosely coupled.